### PR TITLE
Added 'contextDir' configuration option. Used to specify Docker build context

### PR DESCRIPTION
**Pull request by rohanKanojia**
This is the continuation of #995

### CODE REVIEW

**Comment by rohanKanojia** [comment]
This is still wip, I haven't tested it properly and haven't updated docs yet.

File: src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java [hunk 1]
```diff
@@ -130,7 +130,9 @@ public File createDockerTarArchive(String imageName, final MojoParameters params
                 archiveCustomizers.add(new ArchiverCustomizer() {
                     @Override
                     public TarArchiver customize(TarArchiver archiver) throws IOException {
-                        DefaultFileSet fileSet = DefaultFileSet.fileSet(dockerFile.getParentFile());
+                        DefaultFileSet fileSet = DefaultFileSet.fileSet(buildConfig.getContextDir() != null
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java]: According to your change below, `buildConfig.getContextDir()` can never be null.

File: src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java [hunk 2]
```diff
@@ -548,20 +576,48 @@ private void initDockerFileFile(Logger log) {
...
             File dFile = new File(dockerFile);
-            if (dockerFileDir == null) {
+            if (dockerFileDir == null && contextDir == null) {
                 return dFile;
             } else {
-                if (dFile.isAbsolute()) {
-                    throw new IllegalArgumentException("<dockerFile> can not be absolute path if <dockerFileDir> also set.");
+                if(contextDir != null) {
+                    if(dockerFileDir != null) {
+                        log.warn("using contextDir as it has more precedence for Dockerfile");
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java]: Precedence over what ? Suggestion: "Both contextDir (%s) and deprecated dockerFileDir (%s) are configured. Using contextDir." (with inserting the concrete values)

File: src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java [hunk 3]
```diff
@@ -548,20 +576,48 @@ private void initDockerFileFile(Logger log) {
...
+                    }
+
+                    if (dFile.isAbsolute()) {
+                        try {
+                            Files.copy(dFile.toPath(), Paths.get(contextDir));
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java]: I don't think that the dockerfile should be copied int the `contextDir` as this is typically a source code controlled directory. Intead it should be copied into the build dir where the context is copied to for creating the tar archive to be send to the Docker daemon.

File: src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java [hunk 4]
```diff
@@ -548,20 +576,48 @@ private void initDockerFileFile(Logger log) {
...
+                    }
+                    log.warn("dockerFileDir parameter is deprecated, please migrate to contextDir");
+                    return new File(dockerFileDir, dockerFile);
                 }
-                return new File(dockerFileDir, dockerFile);
             }
         }
 
+
+        if (contextDir != null) {
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java]: I recommend to not duplicate that check but just add it outside the conditionals (and use the comment as mentioned above)

File: src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java (hunk 3 above)
Comment (rohanKanojia) [reply src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java]: Could you please tell me where are you creating tar archive? Is it here:

https://github.com/fabric8io/docker-maven-plugin/blob/dc82c82360afca7451b98d89f046204b506bec1e/src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java#L130-L140

Comment (rhuss) [reply src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java]: It's here where the Dockerfile is copied to its target directory: https://github.com/fabric8io/docker-maven-plugin/blob/dec36f1b75f86fa462e125172a183b7355ec460c/src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java#L196-L202

Comment (rohanKanojia) [reply src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java]: okay, thanks

Comment (rohanKanojia) [reply src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java]: to bo honest, I am still confused. Aren't we copying `Dockerfile` here already into build directory? Could you please tell me also where is `dockerFileDir`/`contextDir` getting copied? do we need to copy `Dockerfile` just inside build or inside the `contextDir`?

Comment (rohanKanojia) [reply src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java]: hey, we were not done with this comment? Is it okay to merge it without addressing it?

**Pull request merged**